pkg/wal/wal.go

```diff
@@ -42,19 +75,759 @@ type Segment interface {
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int64(entryLen)); err != nil {
+			return errors.Wrap(err, "Write entry length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int16(seriesIDBytesLen)); err != nil {
+			return errors.Wrap(err, "Write seriesID length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, seriesIDBytes); err != nil {
+			return errors.Wrap(err, "Write seriesID error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int32(len(timestamps))); err != nil {
+			return errors.Wrap(err, "Write series count error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int16(timestampsBytesLen)); err != nil {
+			return errors.Wrap(err, "Write timestamps length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, timestampsBytes); err != nil {
+			return errors.Wrap(err, "Write timestamps error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, valuesBytes); err != nil {
+			return errors.Wrap(err, "Write values error")
+		}
+	}
+	// Rewrite batch length
+	batchBytes := log.bytesBuffer.Bytes()
+	batchLen = len(batchBytes) - batchLength
+	rewriteInt64InBuf(batchBytes, int64(batchLen), 0, binary.LittleEndian)
+
+	return log.writeWorkSegment(batchBytes)
+}
+
+func (log *log) swapWorkSegment() (Segment, error) {
+	log.workSegmentMutex.Lock()
+	defer log.workSegmentMutex.Unlock()
+
+	if err := log.workSegment.file.Close(); err != nil {
+		return nil, errors.Wrap(err, "Close WAL segment error")
+	}
+
+	oldSegment := log.workSegment
+	// Create new segment.
+	segmentID := log.workSegment.segmentID + 1
+	segment := &segment{
+		segmentID: segmentID,
+		path:      filepath.Join(log.path, segmentName(uint64(segmentID))),
+	}
+	var err error
+	segment.file, err = os.OpenFile(segment.path, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.ModePerm)
+	if err != nil {
+		return nil, errors.Wrap(err, "Open WAL segment error")
+	}
+	log.workSegment = segment
+
+	return oldSegment, nil
+}
+
+func (log *log) writeWorkSegment(data []byte) error {
+	log.workSegmentMutex.Lock()
+	defer log.workSegmentMutex.Unlock()
+
+	// Write batch data to WAL segment file
+	if _, err := log.workSegment.file.Write(data); err != nil {
+		return errors.Wrap(err, "Write WAL segment file error, file: "+log.workSegment.path)
+	}
+	if err := log.workSegment.file.Sync(); err != nil {
+		log.logger.Warn().Msg("Sync WAL segment file to disk failed, file: " + log.workSegment.path)
+	}
+	return nil
+}
+
+func (log *log) load() error {
+	files, err := os.ReadDir(log.path)
+	if err != nil {
+		return errors.Wrap(err, "Can not read dir: "+log.path)
+	}
+	// Load all of WAL segments.
+	var workSegmentID SegmentID
+	log.segmentMap = make(map[SegmentID]*segment)
+	for _, file := range files {
+		name := file.Name()
+		segmentID, parsePathErr := strconv.ParseUint(name[3:19], 10, 64)
```

Comment:
It's better to create `parseSegmentID` next to `segmentName`. Then add some method comments to explain the format of the segment name.